pootle/core/state.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class ItemState(object):
+
+    def __init__(self, state, state_type, **kwargs):
+        self.state = state
+        self.state_type = state_type
+        self.kwargs = kwargs
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, self.__class__)
+            and other.state_type == self.state_type
+            and other.state == self.state
+            and (sorted(self.kwargs.items())
+                 == sorted(other.kwargs.items())))
+
+    def __str__(self):
+        return (
+            "<%s(%s): %s %s>"
+            % (self.__class__.__name__,
+               self.state.context,
+               self.state_type,
+               self.kwargs))
+
+    def __getattr__(self, k):
+        if k in self.kwargs:
+            return self.kwargs[k]
+        return self.__getattribute__(k)
+
+
+class State(object):
+
+    item_state_class = ItemState
+    prefix = "state"
+
+    def __init__(self, context, **kwargs):
+        self.context = context
+        self.__state__ = {}
+        self.kwargs = kwargs
+        self.reload()
+
+    def __eq__(self, other):
+        return (
+            self.__class__ == other.__class__
+            and self.context == other.context
+            and sorted(self.kwargs.items()) == sorted(other.kwargs.items())
+            and self.prefix == other.prefix)
+
+    def __contains__(self, k):
+        return k in self.__state__ and self.__state__[k]
+
+    def __getitem__(self, k):
+        return self.__state__[k]
+
+    def __iter__(self):
+        for k in self.__state__:
+            if self.__state__[k]:
+                yield k
+
+    def __str__(self):
+        if not self.has_changed:
+            return (
+                "<%s(%s): Everything up-to-date>"
+                % (self.__class__.__name__, self.context))
+        return (
+            "<%s(%s): %s>"
+            % (self.__class__.__name__,
+               self.context,
+               ', '.join(["%s: %s" % (k, len(self.__state__[k]))
+                          for k in self.states
+                          if self.__state__[k]])))
+
+    @property
+    def has_changed(self):
+        return any(self.__state__.values())
+
+    def add(self, k, v):
+        if k in self.__state__:
+            self.__state__[k].append(v)
+
+    @property
+    def states(self):
```

Comment:
Please put the properties together, following convention for order in models (despite this is not a model) http://docs.translatehouse.org/projects/pootle/en/latest/developers/styleguide.html#order-in-models